File: Scripts/ShaderCompiler.js
const path = require('path');
const fs = require('fs');
const child_process = require('child_process');

/*
    Argument list
    -dxlevel: 9, 11, 12 | DirectX version.
    -olevel 0, 1, 2, 3  | Optimization level for DX9 and DX11.
*/

const directoryPath = path.join(__dirname, 'Shaders');
const args = process.argv.slice(2);

const level = args.findIndex((arg)=> arg === "-dxlevel");
const optim = args.findIndex((arg)=> arg === "-olevel");

console.log(`[INFO]: Compiling shaders...`);

fs.readdir(directoryPath, (err, files) => {
    if (err)
        return console.error(`[FATAL]: Unable to scan directory: ${err}`);

    files.forEach((file) => {
        const fileExt = file.split(".");

        if (fileExt[fileExt.length-1] === "def")
            return;

        if (fileExt[fileExt.length-1] === "frag" || fileExt[fileExt.length-1] === "vert") {
            console.info(`[INFO]: "${file}" is being compiled...`);
            compileShader(file);
            return;
        }

        console.warn(`[WARN]: "${fileExt[1]}" is not a supported format`);
    });
});

function compileShader(file) {
    const name = file.split(".");

    const input = `-f ./Shaders/${file}`;
    const output = `-o ./DX9/${file.split(".")[0]}.${name[1] == "vert" ? "bvs" : "bfs"}`;
    const type = `--type ${name[1] == "vert" ? "vertex" : "fragment"}`;
    const include = "-i ./";
    const platform = "--platform windows";
    const varying = "--varyingdef ./Shaders/Varying.def";
    const profile = "-p s_3_0"; // dx9=s_3_0 dx10=s_4_0 dx11+=s_5_0
    const optlevel = `-O ${optim != -1 ? args[optim+1] : ""}`;
    const defines = "";

    // every array item is need to be seperated at spaces...
    const array = [input, output, type, include, platform, varying, profile, optlevel].join(" ").split(" ");
    const child = child_process.spawn("./Binaries/Win32/shaderc", array);

    child.stdout.setEncoding('utf8');
    child.stdout.on('data', function (data) {
        var str = data.toString()
        var lines = str.split(/(\r?\n)/g);
        console.log(lines.join(""));
    });
}

File: Scripts/ShaderCompiler.py
import os
import sys
import subprocess

# TODO: implement unix timestamping of shit loool :D
# TODO: from bgfx_shader.sh remove the predefined uniforms at the bottom

"""""
    Argument list
    -dxlevel | 9, 11, 12  | DirectX version.                   | Defaults to 9
    -olevel  | 0, 1, 2, 3 | Optimization level. (DX9 and DX11) | Defaults to 0
"""

# Arguments
argc = len(sys.argv)
argv = sys.argv[1:]

# Path definitions
in_folder: str = "./Shaders/"
out_folder: str = "./D3D9/" # TODO: remove hard coded
compiler_path: str = "./Binaries/Win32/shaderc"
varying_path: str = "./Shaders/Varying.def"
include_path: str = "./Shaders/Include/"

# Apply arguments
#opt_dxlevel = argv.index("-dxlevel")
#out_folder = "./D3D" + argv[argv.index("-dxlevel") + 1]

# Functions
def compile_vertex(shader: str):
    in_path: str = f"-f {in_folder}{shader}"
    out_path: str = f"-o {out_folder}{shader}".replace(".vert", ".bvs")
    type: str = "--type vertex"
    include: str = f"-i {include_path}"
    platform: str = "--platform windows"
    varying: str = f"--varyingdef {varying_path}"
    profile: str = "-p s_3_0" # TODO: remove hard coded profile version
    optlevel: str = "-O 3" # TODO: remove hard coded optimization level

    command = [compiler_path, in_path, out_path, type, include, platform, varying, profile, optlevel]
    exitcode = subprocess.run(" ".join(command).split(" ")).returncode

    exitcode or print(f"Compiled {shader}")
    exitcode and print(f"Failed {shader}")

def compile_fragment(shader: str):   
    in_path: str = f"-f {in_folder}{shader}"
    out_path: str = f"-o {out_folder}{shader}".replace(".frag", ".fvs")
    type: str = "--type fragment"
    include: str = f"-i {include_path}"
    platform: str = "--platform windows"
    varying: str = f"--varyingdef {varying_path}"
    profile: str = "-p s_3_0" # TODO: remove hard coded profile version
    optlevel: str = "-O 3" # TODO: remove hard coded optimization level

    command = [compiler_path, in_path, out_path, type, include, platform, varying, profile, optlevel]
    exitcode = subprocess.run(" ".join(command).split(" ")).returncode

    exitcode or print(f"Compiled {shader}")
    exitcode and print(f"Failed {shader}")

# Main
print("\n--- Shader Compiler Script ---\n")

for f in os.listdir(in_folder):
    if f.endswith(".frag"):
        compile_fragment(f)
        continue

    if f.endswith(".vert"):
        compile_vertex(f)